lib/Platform/Intl/PlatformIntlApple.mm

```diff
@@ -24,6 +24,146 @@
+    // decrease pos by 2.
+    if (pos >= 2 && candidate[pos - 2] == '-') {
+      pos -= 2;
+    }
+    // d. Let candidate be the substring of candidate from position 0,
+    // inclusive, to position pos, exclusive.
+    candidate.resize(pos);
+  }
+}
+
+// Implementer note: For more information review
+// https://402.ecma-international.org/7.0/#sec-unicode-locale-extension-sequences
+std::u16string toNoExtensionsLocale(const std::u16string &locale) {
+  std::vector<std::u16string> subtags;
+  auto s = locale.begin(), e = locale.end();
+  while (true) {
+    auto tagEnd = std::find(s, e, u'-');
+    subtags.emplace_back(s, tagEnd);
+    if (tagEnd == e)
+      break;
+    s = tagEnd + 1;
+  }
+  std::u16string result;
+  size_t size = subtags.size();
+  for (size_t s = 0; s < size;) {
+    result.append(subtags[s]);
+    s++;
+    // If next tag is a private marker and there are remaining tags
+    if (subtags[s] == u"u" && s < size - 1)
+      // Skip those tags until you reach end or another singleton subtag
+      while (s < size && subtags[s].size() > 1)
+        s++;
+  }
+  return result;
+}
+
+// Implementer note: This method corresponds roughly to
+// https://tc39.es/ecma402/#sec-lookupmatcher
+struct LocaleMatch {
+  std::u16string locale;
+  std::u16string extension;
+};
+LocaleMatch lookupMatcher(
+    std::vector<std::u16string> &requestedLocales,
+    std::vector<std::u16string> &availableLocales) {
+  // 1. Let result be a new Record.
+  LocaleMatch result;
+  // 2. For each element locale of requestedLocales, do
+  for (std::u16string locale : requestedLocales) {
+    // a. Let noExtensionsLocale be the String value that is locale with
+    // any Unicode locale extension sequences removed.
+    std::u16string noExtensionsLocale = toNoExtensionsLocale(locale);
+    // b. Let availableLocale be BestAvailableLocale(availableLocales,
+    // noExtensionsLocale).
+    llvh::Optional<std::u16string> availableLocale =
+        bestAvailableLocale(availableLocales, noExtensionsLocale);
+    // c. If availableLocale is not undefined, then
+    if (availableLocale) {
+      // i. Set result.[[locale]] to availableLocale.
+        result.locale = std::move(*availableLocale);
+      // ii. If locale and noExtensionsLocale are not the same String value,
+      if (locale != noExtensionsLocale) {
+        // then
+        // 1. Let extension be the String value consisting of the substring of
+        // the Unicode locale extension sequence within locale.
+        // 2. Set result.[[extension]] to extension.
+        result.extension = result.locale.substr(noExtensionsLocale.length(), locale.length());
+      }
+      // iii. Return result.
+      return result;
+    }
+  }
+  // 3. Let defLocale be DefaultLocale().
+  NSString *defLocale = [[NSLocale currentLocale] localeIdentifier];
+  // 4. Set result.[[locale]] to defLocale.
+  result.locale = nsStringToU16String(defLocale);
```

Comment:
Let's also factor out the code to get the current default locale identifier as a `u16string`, then there is nothing platform specific about this function.